### PR TITLE
[FIX] website_livechat: access error when retrieving chatter info

### PR DESCRIPTION
**Pull request by luisg123v**
When chatter information is retrieved but current user has no access to website visitors, an access error is raised and chatter messages are not displayed.

To replicate:
- Enable livechat for current website
- Set user A as member of the livechat
- Log-in with user B, who doesn't have access to website visitors
- Visit any page that automatically opens the livechat with user B (no
  need to input anything, as a livechat is created when the chat window
  pops up)
- Go to backend with user B

To solve the above, visitor information is included only if current user
has enough right to access it.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

OPW #2743176.

### CODE REVIEW

**Comment by luisg123v** [comment]
Hi @moylop260,

What do you think?

CC @hugho-ad

**Comment by moylop260** [comment]
@mart-e 

Is it should be fixed with ACL or return early?

**Comment by mart-e** [comment]
return early seems a better idea than changing the access rights. You could also use a try/expect AccessError to avoid an additional check when it's not needed

File: addons/website_livechat/models/mail_channel.py [hunk 1]
```diff
@@ -28,6 +28,8 @@ def channel_info(self, extra_info=False):
         at the top of the livechat channel discussion view in discuss module.
         """
         channel_infos = super(MailChannel, self).channel_info(extra_info)
+        if not self.env['website.visitor'].check_access_rights('read', raise_exception=False):
```
Comment (seb-odoo) [inline addons/website_livechat/models/mail_channel.py:31]: Also if you keep the "check first" version, you would need to also call `check_access_rule` (on principle at least, I'm not sure whether visitor actually has specific rules that would make this an issue).

**Comment by luisg123v** [comment]
Hi @mart-e & @seb-odoo,

Done, changed by a try-except.

Regarding record rules, there's currently no rules for website visitors. However, I tested creating a new one and it still worked when catching `AccessError`.

Thanks a lot!

Could you review again, please?

CC @CarmenMiranda

File: addons/website_livechat/models/mail_channel.py [hunk 2]
```diff
@@ -30,20 +31,29 @@ def channel_info(self, extra_info=False):
...
-                    'website': visitor.website_id.name,
-                    'lang': visitor.lang_id.name,
-                    'partner_id': visitor.partner_id.id,
-                }
+            try:
+                visitor_info = channel._get_livechat_visitor_info()
+            except AccessError:
+                visitor_info = {}
+            if visitor_info:
+                channel_infos_dict[channel.id]['visitor'] = visitor_info
```
Comment (seb-odoo) [inline addons/website_livechat/models/mail_channel.py:39]: Move this into the try block as a single line instead
We don't want to add an empty value in case of exception, because that indicates there is no visitor (when it's not actually true). It's better to not add the key to indicate there is no data.

File: addons/website_livechat/models/mail_channel.py [hunk 3]
```diff
@@ -30,20 +31,29 @@ def channel_info(self, extra_info=False):
...
         return list(channel_infos_dict.values())
 
+    def _get_livechat_visitor_info(self):
```
Comment (seb-odoo) [inline addons/website_livechat/models/mail_channel.py:42]: This is too close to `_channel_get_livechat_visitor_info` and should probably be refactored in master. I would be against introducing this new method here for this reason, and keep the code were it was and avoid unnecessary diff.

**Review by seb-odoo — CHANGES_REQUESTED** [review]

File: addons/website_livechat/models/mail_channel.py (hunk 2 above)
Comment (luisg123v) [reply addons/website_livechat/models/mail_channel.py]: Done, moved. However, an empty value wasn't going to be added thanks to the conditional `if visitor_info` just above.

File: addons/website_livechat/models/mail_channel.py (hunk 3 above)
Comment (seb-odoo) [reply addons/website_livechat/models/mail_channel.py]: If you want to avoid indentation level (which I think is a good idea too) and reduce diff to absolute minimum, what you can do is do "if not visitor: continue", then add the `try:` for which the rest of the code won't change indentation.

File: addons/website_livechat/models/mail_channel.py [hunk 4]
```diff
@@ -32,16 +33,19 @@ def channel_info(self, extra_info=False):
...
+                        'country_code': visitor.country_id.code.lower() if visitor.country_id else False,
+                        'country_id': visitor.country_id.id,
+                        'is_connected': visitor.is_connected,
+                        'history': self.sudo()._get_visitor_history(visitor),
+                        'website': visitor.website_id.name,
+                        'lang': visitor.lang_id.name,
+                        'partner_id': visitor.partner_id.id,
+                    }
+                except AccessError:
+                    continue
```
Comment (seb-odoo) [inline addons/website_livechat/models/mail_channel.py:48]: `pass` would be more clear here, as we really want do "do nothing" (`continue` wouldn't work if we had/need to have code after this try/catch block)

File: addons/website_livechat/models/mail_channel.py (hunk 3 above)
Comment (seb-odoo) [reply addons/website_livechat/models/mail_channel.py]: Alternatively (and I think it's more usually done in python but I'm not sure), it's to pre-filter the loop variable so that the `if` is not needed at all.

**Comment by luisg123v** [comment]
@seb-odoo done, test case was added.

It checks the key is added when current user has permissions, and it's not added when the user does not.

I also ran the test without these changes and confirmed it crashes without them.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
